itest/rust/src/dictionary_test.rs

```diff
@@ -400,3 +425,260 @@ fn dictionary_equal() {
+                i64::try_from_variant(&v).unwrap(),
+            )
+        })
+        .collect();
+    assert_eq!(map, collected_map);
+    let collected_dictionary: Dictionary = collected_map.into_iter().collect();
+    assert_eq!(dictionary, collected_dictionary);
+}
+
+// The below tests, test to ensure no panicking happens when mutating
+// a dictionary while iterating. Though the result of doing is not
+// guaranteed.
+
+#[itest]
+fn dictionary_iter_clear() {
+    let dictionary = dict! {
+        "foo": 0,
+        "bar": true,
+        "baz": "foobar",
+        "nil": Variant::nil(),
+    };
+    let mut dictionary2 = dictionary.share();
+
+    let mut iter = dictionary.iter_shared();
+    iter.next();
+    iter.next();
+    dictionary2.clear();
+    let _v: Vec<_> = iter.collect();
+}
+
+#[itest]
+fn dictionary_iter_insert() {
+    let dictionary = dict! {
+        "foo": 0,
+        "bar": true,
+        "baz": "foobar",
+        "nil": Variant::nil(),
+    };
+    let mut dictionary2 = dictionary.share();
+
+    let mut iter = dictionary.iter_shared();
+    iter.next();
+    iter.next();
+
+    dictionary2.insert("new_key", 10);
+    let _v: Vec<_> = iter.collect();
+}
+
+#[itest]
+fn dictionary_iter_insert_after_completion() {
+    let dictionary = dict! {
+        "foo": 0,
+        "bar": true,
+        "baz": "foobar",
+        "nil": Variant::nil(),
+    };
+    let mut dictionary2 = dictionary.share();
+    let mut iter = dictionary.iter_shared();
+    for _ in 0..4 {
+        iter.next();
+    }
+    assert_eq!(iter.next(), None);
+
+    dictionary2.insert("new_key", 10);
+    iter.next();
+}
+
+#[itest]
+fn dictionary_iter_big() {
+    let dictionary: Dictionary = (0..256).zip(0..256).collect();
+    let mut dictionary2 = dictionary.share();
+    let mut iter = dictionary.iter_shared();
+
+    for _ in 0..4 {
+        for _ in 0..4 {
+            for _ in 0..16 {
+                iter.next();
+            }
+            dictionary2.insert("a", "b");
+        }
+        dictionary2.clear();
+        dictionary2.extend((0..64).zip(0..64));
+    }
+}
+
+#[itest]
+fn dictionary_iter_clear_obj() {
+    let dictionary = dict! {
+        (user_object()): user_object(),
+        "foo": 10,
+        (user_object()): user_object(),
+    };
+    let mut dictionary2 = dictionary.share();
+
+    let mut iter = dictionary.iter_shared();
+
+    iter.next();
+    dictionary2.clear();
+    let _v: Vec<_> = iter.collect();
+}
+
+#[itest]
+fn dictionary_iter_twice() {
+    let dictionary = dict! {
+        "foo": 10,
+        "bar": true,
+        "baz": "foobar",
+        "nil": Variant::nil(),
+    };
+
+    let map: HashMap<String, (Variant, Variant)> = dictionary
+        .iter_shared()
+        .zip(dictionary.iter_shared())
+        .map(|((k1, v1), (k2, v2))| {
+            let mut s1 = String::try_from_variant(&k1).unwrap();
+            let s2 = String::try_from_variant(&k2).unwrap();
```

Comment:
`from_variant`

Reply:
i think i initially didn't use `from_variant` because i initially saw it was behind a `#[cfg(..)]` thing and so stayed away from it in library code. but i can do what you suggest if we're not gonna keep that.

Reply:
I see. Yeah, I'll likely remove that feature.